add error mitigation explanation notebook

kevinsung opened the pull request:
Fixes #1355 .

No material on PEA yet because it's not exposed as an [option](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/options).

kevinsung commented:
Hmm the embedded images seem to be giving the CI issue. Gonna need to investigate...

javabster commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,274 @@
+    }
+   },
+   "source": [
+    "# Error mitigation technique explanations\n",
```
Hmm. I think we can come up with a better title 🤔 

How about just "Error mitigation techniques". Or "Error mitigation techniques available in Qiskit Runtime"? or "Qiskit Runtime error mitigation techniques"

kevinsung replied:
Went with "Error mitigation techniques"

javabster commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "Quantum circuits are executed on IBM hardware as sequences of microwave pulses that need to be scheduled and run at precise time intervals.\n",
+    "Unfortunately, unwanted interactions between qubits can lead to coherent errors on idling qubits. Dynamical decoupling works by inserting pulse sequences on idling qubits to approximately cancel out the effect of these errors. Each inserted pulse sequence amounts to an identity operation, but the physical presence of the pulses has the effect of suppressing errors.\n",
+    "There are many possible cohoice of pulse sequences, and which sequence is better for each particular case remains an [active area of research](https://journals.aps.org/prapplied/abstract/10.1103/PhysRevApplied.20.064027).\n",
```
```suggestion
    "There are many possible choices of pulse sequences, and which sequence is better for each particular case remains an [active area of research](https://journals.aps.org/prapplied/abstract/10.1103/PhysRevApplied.20.064027).\n",
```

kevinsung replied:
Done.

javabster commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,274 @@
+   "source": [
+    "estimator = Estimator(backend=backend)\n",
+    "estimator.options.dynamical_decoupling.enable = True\n",
+    "estimator.options.dynamical_decoupling.sequence_type = \"XpXm\""
```
is setting the sequence type optional or required? I think it would be good to mention either way

javabster replied:
Also I notice in the other sections you explicitly say that to enable the technique you must set enable = true but you dont say that here. I think you should either say that point for all of them or none of them, but just make sure it's consistent

javabster commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,274 @@
+   "outputs": [],
+   "source": [
+    "estimator = Estimator(backend=backend)\n",
+    "estimator.options.twirling.enable_gates = True\n",
```
same with these options, it would be good to call out to the user which of these options you have to set vs which are optional

kevinsung replied:
Done.

javabster commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,274 @@
+    "\n",
+    "In the presence of readout error, measurement twirling has the effect of diagonalizing the readout-error transfer matrix, making it easy to invert. Estimating the readout-error transfer matrix requires executing additional calibration circuits, which introduces a small overhead.\n",
+    "\n",
+    "TREX can be enabled by setting `measure_mitigation` to `True` in version 2 of the [Qiskit Runtime resilience options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2). Options for measurement twirling are described [here](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.MeasureNoiseLearningOptions). As in gate t ...
```
```suggestion
    "TREX can be enabled by setting `measure_mitigation` to `True` in version 2 of the [Qiskit Runtime resilience options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2). Options for measurement twirling are described [here](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.MeasureNoiseLearningOptions). As with gate twirling, you can set the number of circuit randomizations and the number of shots per randomization.\n",
```

I think this is slightly better grammar, but @abbycross will know better 😅

kevinsung replied:
Done.

javabster commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,274 @@
+   "outputs": [],
+   "source": [
+    "estimator = Estimator(backend=backend)\n",
+    "estimator.options.resilience.measure_mitigation = True\n",
```
again can you mention which is optional and which are not pls :)

kevinsung replied:
Done.

javabster reviewed (commented):
Looking really good! I think this is going to be a very helpful docs page.

I left some in line comments and I have these general ones:

- please add a "next steps" section like we do for the other docs pages. I think we can add a link to the tutorial on combining EM techniques, as well as maybe the API reference
- Do you think we should add a section about PEA? I know it's technically only available through the experimental branch for now but we have docs for other experimental features so I can see a case for including here too

kevinsung commented:
@javabster I've addressed your comments.

>Do you think we should add a section about PEA? I know it's technically only available through the experimental branch for now but we have docs for other experimental features so I can see a case for including here too

I think we should document PEA once it's available, since the way that the options are exposed would probably affect the exposition. We shouldn't document options that aren't available yet.

javabster commented:
@jyu00 would you mind giving this a review when you get a moment?

jyu00 commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+    "\n",
+    "In the presence of readout error, measurement twirling has the effect of diagonalizing the readout-error transfer matrix, making it easy to invert. Estimating the readout-error transfer matrix requires executing additional calibration circuits, which introduces a small overhead.\n",
+    "\n",
+    "TREX can be enabled by setting `measure_mitigation` to `True` in version 2 of the [Qiskit Runtime resilience options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2). Options for measurement twirling are described [here](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.MeasureNoiseLearningOptions). As with gate ...
```
1. The class is called `ResilienceOptionsV2` because it's for V2 Estimator. Saying it's "version 2 of resilience options" might make people think you can use v1 options with v2 primitives. 
2. Since the link points to `MeasureNoiseLearningOptions` (same as the code), I assume that's what you're referring to. 


```suggestion
    "TREX can be enabled by setting `measure_mitigation` to `True` in the [Qiskit Runtime resilience options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2) for V2 Estimator. Options for measurement noise learning are described [here](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.MeasureNoiseLearningOptions). As with gate twirling, you can set the number of circuit randomizations and the number of shots per randomization.\n",
```

jyu00 commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+    "\n",
+    "![Depiction of ZNE](/images/run/error-mitigation-explanation/zne.png)\n",
+    "\n",
+    "ZNE can be enabled by setting `zne_mitigation` to `True` in version 2 of the [Qiskit Runtime resilience options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2).\n",
```
```suggestion
    "ZNE can be enabled by setting `zne_mitigation` to `True` in the [Qiskit Runtime resilience options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2) for V2 Estimator.\n",
```

jyu00 reviewed (approved):
Some minor comments but I think this looks great!

abbycross commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+   "source": [
+    "## Dynamical decoupling\n",
+    "\n",
+    "Quantum circuits are executed on IBM hardware as sequences of microwave pulses that need to be scheduled and run at precise time intervals.\n",
```
```suggestion
    "Quantum circuits are executed on IBM&reg; hardware as sequences of microwave pulses that need to be scheduled and run at precise time intervals.\n",
```

abbycross commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+    "![Depiction of ZNE](/images/run/error-mitigation-explanation/zne.png)\n",
+    "\n",
+    "ZNE can be enabled by setting `zne_mitigation` to `True` in version 2 of the [Qiskit Runtime resilience options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2).\n",
+    "The Qiskit Runtime options for ZNE are described [here](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ZneOptions). The following options are notable:\n",
```
```suggestion
    "The Qiskit Runtime options for ZNE are described [here](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ZneOptions). The following options are notable:\n",
```

abbycross commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+    "- `noise_factors`: The noise factors to use for noise amplification.\n",
+    "- `extrapolator`: The functional form to use for the extrapolation.\n",
+    "\n",
+    "The following code cell shows how enable ZNE and set these options for the estimator. None of these options are required to be set explicitly."
```
```suggestion
    "The following code cell shows how to enable ZNE and set these options for the estimator. None of these options are required to be set explicitly."
```

abbycross commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+    "\\mathcal{O}_{\\text{ideal}} = \\sum_{i} \\eta_i \\mathcal{O}_{noisy, i}\n",
+    "$$\n",
+    "\n",
+    "The output of the ideal circuit can then be reproduced by executing different noisy circuit instances drawn from a random ensemble defined by the linear combination. If the coefficients $\\eta_i$ form a probability distribution, they can be used directly as the probabilities of the ensemble. In practice, some of the coefficients are negative, so they form a quasiprobability distribution inst ...
```
```suggestion
    "The output of the ideal circuit can then be reproduced by executing different noisy circuit instances drawn from a random ensemble defined by the linear combination. If the coefficients $\\eta_i$ form a probability distribution, they can be used directly as the probabilities of the ensemble. In practice, some of the coefficients are negative, so they form a quasiprobability distribution instead. They can still be used to define a random ensemble, but there is a sampling overhead related to the negativity of the quasi-probability distribution, which is characterized by the quantity\n",
```

kevinsung replied:
What's the reason for this suggestion?

abbycross replied:
I knew you'd ask about this 🤣 The consensus at IBM Quantum is to use `quasi-probability distribution` as the standard. https://github.com/Qiskit/qiskit-ibm-runtime/pull/924

abbycross commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+    "The sampling overhead is a multiplicative factor on the number of shots required to estimate an expectation value to a given precision, compared to the number of shots that would be needed from the ideal circuit. It scales quadratically with $\\gamma$, which in turn scales exponentially with the depth of the circuit.\n",
+    "\n",
+    "ZNE can be enabled by setting `pec_mitigation` to `True` in version 2 of the [Qiskit Runtime resilience options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.ResilienceOptionsV2).\n",
+    "The Qiskit Runtime options for PEC are described [here](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.PecOptions). A limit on the sampling overhead can be set using the `max_overhead` option. Note that limiting the sampling overhead can cause the precision of the result to exceed the requested precision. The default value of `max_overhead` is 100.\n",
```
```suggestion
    "The Qiskit Runtime options for PEC are described [here](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.PecOptions). A limit on the sampling overhead can be set using the `max_overhead` option. Note that limiting the sampling overhead can cause the precision of the result to exceed the requested precision. The default value of `max_overhead` is 100.\n",
```

abbycross commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+    "## Next steps\n",
+    "\n",
+    "- Check out the [tutorial](https://learning.quantum.ibm.com/tutorial/combine-error-mitigation-options-with-the-estimator-primitive) on combining error mitigation options with the Estimator primitive.\n",
+    "- Visit the [API reference](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/options) to see all the options for the Qiskit Runtime primitives."
```
```suggestion
    "- Visit the [API reference](/api/qiskit-ibm-runtime/options) to see all the options for the Qiskit Runtime primitives."
```

abbycross reviewed (approved):
I caught a few small word things - otherwise it reads very well, as always!

kevinsung commented on docs/run/error-mitigation-explanation.ipynb:
```diff
@@ -0,0 +1,284 @@
+    "Dynamical decoupling can be enabled by setting `enable` to `True` in the [dynamical decoupling options](https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.options.DynamicalDecouplingOptions). The `sequence_type` option can be used to pick from several different pulse sequences. The default sequence type is `\"XX\"`.\n",
+    "\n",
+    "The following code cell shows how to enable dynamical decoupling for the estimator and choose a dynamical decoupling sequence."
+   ]
```
I added this paragraph:
>Note that dynamical decoupling is mainly useful for circuits containing gaps in which some qubits sit idle without any operations acting on them. If the operations in the circuit are packed very densely, such that all of the qubits are busy most of the time, then the addition of dynamical decoupling pulses might not improve performance. In fact, it could even worsen performance due to imperfections in the pulses themselves.
